### PR TITLE
feat:[CDS-31886]: notification for approval or rejection action in approval step

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
develop PR for the same ticket 
https://github.com/harness/harness-core/pull/47890
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47978)

### CODE REVIEW

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 13 of 14 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47978)* status: 1 file(s) to review.

File: pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java [hunk 1]
```diff
@@ -297,6 +326,80 @@ private NotificationChannel getNotificationChannel(HarnessApprovalInstance insta
...
+  private String getAction(HarnessApprovalInstance approvalInstance) {
+    Optional<HarnessApprovalActivity> optionalHarnessApprovalActivity = approvalInstance.fetchLastApprovalActivity();
+    String action = "";
+    if (optionalHarnessApprovalActivity.isPresent()) {
+      HarnessApprovalActivity lastApprovalActivity = optionalHarnessApprovalActivity.get();
+      if (HarnessApprovalAction.APPROVE.equals(lastApprovalActivity.getAction())) {
+        List<HarnessApprovalActivity> harnessApprovalActivities = approvalInstance.getApprovalActivities();
+        for (HarnessApprovalActivity harnessApprovalActivity : harnessApprovalActivities) {
+          action = action
+              + (harnessApprovalActivity.getUser().getEmail() + " approved on "
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:387]: Need to default to something here when email is not present, please check getUser method.

File: pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java [hunk 2]
```diff
@@ -111,7 +113,17 @@ public ApprovalNotificationHandlerImpl(@Named("PRIVILEGED") UserGroupClient user
...
+      boolean isApprovalAction = ApprovalStatus.APPROVED.equals(approvalInstance.getStatus())
+          || ApprovalStatus.REJECTED.equals(approvalInstance.getStatus());
+      if (isApprovalAction) {
+        ApprovalSummary approvalSummary =
+            getApprovalSummary(ambiance, approvalInstance, getPipelineExecutionSummary(ambiance));
+        sendNotificationInternal(
+            approvalInstance, approvalInstance.getValidatedUserGroups(), approvalSummary.toParams(), logCallback);
+      } else {
+        sendNotificationInternal(approvalInstance, ambiance, logCallback);
+      }
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:126]: nit: we can add a try catch block here for any general exception due to the new logic being added.

File: pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java [hunk 3]
```diff
@@ -297,6 +326,80 @@ private NotificationChannel getNotificationChannel(HarnessApprovalInstance insta
...
+          action = action
+              + (harnessApprovalActivity.getUser().getEmail() + " approved on "
+                  + formatTime(harnessApprovalActivity.getApprovedAt()) + "   \\n");
+        }
+        if (!isEmpty(action)) {
+          // removing last redundant new line character
+          action = action.substring(0, action.length() - 2);
+        }
+      } else if (HarnessApprovalAction.REJECT.equals(lastApprovalActivity.getAction())) {
+        action = lastApprovalActivity.getUser().getEmail() + " rejected on "
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:395]: same user related comment

File: pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java [hunk 4]
```diff
@@ -251,6 +270,17 @@ private NotificationChannel getNotificationChannel(HarnessApprovalInstance insta
       notifyUserGroupBuilder.setProjectIdentifier(userGroup.getProjectIdentifier());
     }
 
+    if (ApprovalStatus.APPROVED.equals(instance.getStatus()) || ApprovalStatus.REJECTED.equals(instance.getStatus())) {
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/notification/ApprovalNotificationHandlerImpl.java:273]: nit : for this expression we can add a utils method in ApprovalStatus enum for readibility

File: pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action.txt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+subject: "Pipeline ${pipelineName} ${status}."
+body: "
+<html>
+<b>Pipeline  <a href=\"${pipelineExecutionLink}\">${pipelineName}</a>  ${status}</b><br><br>
+
+<b>${action}</b><br>
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action.txt:6]: nit: we can add a header for {action} like "Actions"

File: pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action_execution.txt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+<b>Organization:</b> ${orgName}<br>
+<b>Project:</b> ${projectName}<br>
+<b>Triggered by:</b> ${triggeredBy}<br><br>
+
+<b>Pipeline execution summary:</b><br>
+Upcoming Stages: ${upcomingStages}<br>
+Completed Stages: ${finishedStages}<br>
+Running Stages: ${runningStages}<br><br>
+
+<i>Pipeline started on ${startedAt} and ${action} </i><br>
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/resources/notification_templates/approval/email/plain_text_action_execution.txt:17]: Same comment as for normal template

**Review by Jimit09Gandhi — APPROVED** [review]
> Reviewed 13 of 14 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47978)* status: :shipit: complete! ✓

**Comment by Jimit09Gandhi** [comment]
@sourabh-awashti @yogesh-chauhan Is change dependant on PL service as well then HF PL service as well for
platform-service/modules/resource-group-service/contracts/src/main/java/io/harness/resourcegroup/v1/remote/dto/ZendeskHelper.java

**Pull request closed without merging**